cli_tools/gce_export/main.go

```diff
@@ -69,99 +91,247 @@ func splitLicenses(input string) []string {
 	return ls
 }
 
-func main() {
-	flag.Parse()
+type buffer struct {
+	// These fields are read only.
+	cSize    int64
+	prefix   string
+	ctx      context.Context
+	id       string
+	bkt, obj string
 
-	if *localPath != "" && *gcsPath != "" {
-		log.Fatal("-local_path and -gcs_path can't be both specified")
-	}
+	upload    chan string
+	tmpObjs   []string
+	tmpObjsMx sync.Mutex
 
-	if *localPath == "" && *gcsPath == "" {
-		log.Fatal("-local_path or -gcs_path must be provided")
-	}
+	sync.Mutex
+	sync.WaitGroup
+	bytes int64
+	part  int
+	file  *os.File
+}
 
-	if *disk == "" {
-		log.Fatal("The flag -disk must be provided")
+func (b *buffer) addObj(obj string) {
+	b.tmpObjsMx.Lock()
+	b.tmpObjs = append(b.tmpObjs, obj)
+	b.tmpObjsMx.Unlock()
+}
+
+func (b *buffer) uploadWorker() {
+	defer b.Done()
+	for in := range b.upload {
+		for i := 1; ; i++ {
+			err := func() error {
+				client, err := gcsClient(b.ctx)
+				if err != nil {
+					return err
+				}
+				defer client.Close()
+
+				file, err := os.Open(in)
+				if err != nil {
+					return err
+				}
+				defer file.Close()
+				defer os.Remove(in)
+
+				tmpObj := path.Join(b.obj, strings.TrimPrefix(in, b.prefix))
+				b.addObj(tmpObj)
+				dst := client.Bucket(b.bkt).Object(tmpObj).NewWriter(b.ctx)
+				dst.ChunkSize = gcsChunkSize
+				if _, err := io.Copy(dst, file); err != nil {
+					if io.EOF != err {
+						return err
+					}
+				}
+
+				return dst.Close()
+			}()
+			if err != nil {
+				if i > 3 {
+					log.Fatal(err)
+				}
+				time.Sleep(time.Duration(1*i) * time.Second)
```

Comment:
My experience: 3 times (<8s in total) of retry is not enough, especially when we need to wait for the GCS server to recover from a bad status. We can do 2 extra things:
1. retry more, for example 16 times. The total length of retrying period can be 5min.
2. randomize the waiting seconds. For example, int.rand(duration). The reason is that, very likely all uploading threads (from different clients) were cut at the same time. It would be good to avoid all re-connections happened in the same timing.

Due that reset issue has been auto mitigated in recent 4 days, you can leave it to me as a improvement as well.

Reply:
I bumped it to 16, the 3 was fairly arbitrary and a total guess on my part as I have never been able to repo a connection issue. 

As far as randomizing, we could do that but I'm not to worried about a few connection attempts as we default to num of CPUs in order to not cause any contention.

Reply:
"I'm not too worried about a few connection attempts as we default to num of CPUs"
That is true for one instance. However, my assumption is that all export instances connected to that GCS server can be reset at the same time (for example, gcs service got restarted), which can be a lot of connections.